### DISCUSSION TITLE
Heat Map Spectrum to Grayscale (Depth Maps)

### DISCUSSION THREAD

**Discussion by WorldofDepth** [ImageMagick/ImageMagick#3126] [Help]
I would like to do the REVERSE of this, which colorizes a grayscale gradient:
`convert input ( xc:white xc:yellow xc:red xc:purple xc:black +append -filter cubic -resize 600x1! ) -clut output`

—That is, I'd like to take an image like the below, which is a disparity map using a heat map-type color spectrum (as I've estimated above), and convert it to a grayscale gradient depth map. This is not the same, of course, as simply converting a color image to grayscale using luminance or average or other methods.

![46337R-map](https://user-images.githubusercontent.com/77465052/104650080-adb5a800-5683-11eb-855d-8db00e646b44.png)

I have found mention of @Snibgo's "invclut", but I don't understand process modules. I also suspect this could be done with hald cluts, but I don't know how to go about generating the clut.

Any help would be appreciated—thank you!

**Comment by snibgo** [comment]
You have defined a clut image. This transforms a grayscale image to colour. You need the inverse clut that transforms from colour back to the original grayscale. This would need to be a "Hald clut", which is sparsely defined, so is a bit messy.

A simple alternative is to note that when we convert your clut to grayscale, the result decreases monotonically. So each colour maps to a unique grayscale value, and we can use this as the look-up to a simple clut.

That simple clut is your clut converted to grayscale, then inverted (which swaps input and output). My process module "invclut" does this job, but it needs a monotonically _increasing_ input. So we "-flop" before inverting, and "-negate" after inverting. See [Process modules: invert clut](http://im.snibgo.com/customim.htm#invclut)

I resize the clut to 6000 instead of 600 to get better resolution. All commands are Windows BAT syntax.
```
magick ^
  104650080-adb5a800-5683-11eb-855d-8db00e646b44.png ^
  -colorspace Gray ^
  ( xc:white xc:yellow xc:red xc:purple xc:black +repage +append -filter cubic -resize "6000x1^!" ^
    -colorspace Gray ^
    -flop ^
    -process invclut ^
    -negate ^
  ) ^
  -clut ^
  +depth ^
  x.png
```
x.png is the required grayscale image:
![x](https://user-images.githubusercontent.com/33812368/104679192-12edb580-56e5-11eb-9356-09f81e5595be.png)

A round-trip test can check the result by applying your clut to this grayscale image, x.png. We should get your original colour image.
```
magick ^
  x.png ^
  ( xc:white xc:yellow xc:red xc:purple xc:black +append +repage -filter cubic -resize "6000x1^!" ) ^
  -clut ^
  +depth ^
  c.png
```
![c](https://user-images.githubusercontent.com/33812368/104679171-049f9980-56e5-11eb-91a6-b5953630bf19.png)

c.png is very similar to the original colour input, but not identical. (It is more saturated.) I don't know why it is different.

You can build IM with my process modules.

The equivalent of "invclut" can be written as a shell script. It will be slower, but for only 6000 pixels this isn't a huge problem.

**Comment by snibgo** [comment]
I had forgotten that I had devised a method for inverting cluts without process modules or scripts. It works by creating square images, so beware of running out of memory with large cluts, eg larger than 10000x1 pixels.

The results are identical (RMSE 0.009%) to the results I showed above.
```
magick ^
  104650080-adb5a800-5683-11eb-855d-8db00e646b44.png ^
  -colorspace Gray ^
  ( xc:white xc:yellow xc:red xc:purple xc:black +append +repage -filter cubic -resize "6000x1^!" ^
    -colorspace Gray ^
    -flop ^
    -set option:WW %%[fx:w] ^
    -scale "%%[WW]x%%[WW]^!" ^
    -size %%[WW]x%%[WW] gradient: ^
    -compose MinusDst -composite ^
    -channel RGB -threshold 0 +channel ^
    -scale "1x%%[WW]^!" ^
    -rotate 90 ^
    -negate ^
  ) ^
  -clut ^
  +depth ^
  x2.png

magick ^
  x2.png ^
  ( xc:white xc:yellow xc:red xc:purple xc:black +append +repage -filter cubic -resize "6000x1^!" ) ^
  -clut ^
  +depth ^
  c2.png
```

**Reply by WorldofDepth** [reply]
Thank you, @snibgo! I will try out this second method.

The reason the round-trip result is slightly different could be because the original image isn't _exactly_ the combination of colors I estimated—I found out it's actually a colormap scheme by MatLab called Inferno, so it's more continuous:

![inferno](https://user-images.githubusercontent.com/77465052/104791426-9e5c5a80-5768-11eb-8aaf-cc4c57c31572.png)

With your 2nd method using the square, I could convert the spectrum above to a 6000x1 image and use it in place of the parenthetical xc portion of the command, right? (I think there's a close parentheses missing?) Thanks again.

**Comment by snibgo** [comment]
Each command has one pair of parentheses: (). There is no missing close parenthesis.

I'll call that clut image "inferno.png". It is the opposite way round to your previous: black is on the left, not the right, so we don't need a "-negate", and we don't need to make it 6000 pixels wide.
```
magick ^
  104650080-adb5a800-5683-11eb-855d-8db00e646b44.png ^
  -colorspace Gray ^
  ( inferno.png ^
    -colorspace Gray ^
    -set option:WW %%[fx:w] ^
    -scale "%%[WW]x%%[WW]^!" ^
    -size %%[WW]x%%[WW] gradient: ^
    -compose MinusDst -composite ^
    -channel RGB -threshold 0 +channel ^
    -scale "1x%%[WW]^!" ^
    -rotate 90 ^
  ) ^
  -clut ^
  +depth ^
  x4.png
```
![x4](https://user-images.githubusercontent.com/33812368/104795536-15184880-57a7-11eb-9346-4ce11c0ac62f.png)

For the round-trip check, we don't need any parentheses.
```
magick ^
  x4.png ^
  inferno.png ^
  -clut ^
  +depth ^
  c4.png
```
![c4](https://user-images.githubusercontent.com/33812368/104795529-0a5db380-57a7-11eb-9f60-89c590fb2912.png)

The round-trip check is worse than before.

EDIT: inferno.png is 888 pixels wide but has only 256 unique colours. So some (probably all) input grayscale values map to the same output colour. When inverted, some (probably all) input colours do not map to a unique output grayscale.

**Reply by WorldofDepth** [reply]
Thank you, @snibgo! Yes, my mistake misreading the code there—got it now. I just tried this method with other example disparity maps, and indeed it sometimes squashes regions of the spectrum, but works passably well for what I'm trying to do. Thank you for all your thought and help!

**Comment by snibgo** [comment]
The "inferno" colormap image shown above by @WorldofDepth is very similar to (perhaps the same as, except for the repetitions) the numbers given in https://github.com/BIDS/colormap/blob/master/colormaps.py . But neither of those colormaps created the colour image shown in the OP.

To demonstrate this fact, we can find unique colours, order them by increasing intensity, and graph them.

First, the colour image:
```
magick ^
  104650080-adb5a800-5683-11eb-855d-8db00e646b44.png ^
  -unique-colors -process sortpixels ^
  1046_cols.png

call %PICTBAT%graphLineCol 1046_cols.png
```
1046_cols_glc.png:
![1046_cols_glc](https://user-images.githubusercontent.com/33812368/104835329-bd103d80-589d-11eb-8927-29a8dfc4f589.png)

Then the colormap image:
```
magick ^
  infernoX.png ^
  -unique-colors -process sortpixels ^
  infernoX_cols.png

call %PICTBAT%graphLineCol infernoX_cols.png
```
infernoX_cols_glc.png:
![infernoX_cols_glc](https://user-images.githubusercontent.com/33812368/104835340-d5805800-589d-11eb-9687-553c7533fb38.png)

Ignore the x-axis, but examine the actual colours as shown on the y-axis in the two graphs. Where the blue and green lines intersect, the two graphs show intersections at very different y-values. The blue line peaks and troughs also have very different y-values in the two graphs.

In other words, the colour image has colours that are very different to any colours in the colormap.  So that colormap didn't create that colour image.

We can use 1046_cols.png as the colormap (aka "clut file"). This creates the following grayscale:
![x4_1046](https://user-images.githubusercontent.com/33812368/104835353-ee890900-589d-11eb-8748-9f9adf1a5bb0.png)

And we can check the round-trip:
![c4_1046](https://user-images.githubusercontent.com/33812368/104835360-f9dc3480-589d-11eb-9beb-a7d84e8626ad.png)

The round-trip looks like the original colour image, and is accurate (RMSE 0.25%).

I can't claim that 1046_cols.png was the exact clut file used to create the original colour image. The actual clut might have had middle values squished to the left or right. There are an infinite number of colormaps that might have been used.

**Reply by WorldofDepth** [reply]
Ah, very interesting! So perhaps the creators of the original disparity map used something based on but slightly different from Inferno. Anyway, this last method is quite clever, and I think should work for images that represent any  continuous-intensity colormap spectrum without large gaps… I don't suppose there's a way to generate files like the 1046_cols.png without enabling/installing IM process modules, is there?

**Comment by snibgo** [comment]
"-process sortpixels" sorts the pixels in a Nx1 image by intensity. That task could be done in a shell script: write I, R, G, and B to a text file, sort that by I, then read the RGB values.

Throughout, we assume no two colours have the same intensity.

But it depends on what your goal is. An unknown grayscale image was clutted with an unknown colormap to make a _known_ colour image. If we knew either the grayscale or the colormap, we could calculate the other. But with both unknown, we can find a solution but don't know if is the same as the original unknown grayscale image.

We can create another solution that is just as valid by:
```
magick 104650080-adb5a800-5683-11eb-855d-8db00e646b44.png -colorspace Gray g.png
```

**Reply by WorldofDepth** [reply]
Oh yes, I tried this first of all. I know some colormaps are designed to have decent representations in grayscale, but with Inferno, depending on the image, I think some contrast is lost/distorted.

**Comment by snibgo** [comment]
See also https://imagemagick.org/script/command-line-options.php#sort-pixels . This seems to be a recent addition to ImageMagick. Perhaps it does the same as my "-process sortpixels".

**Reply by WorldofDepth** [reply]
Ahh, this would be great! Apparently added 5 days ago. Unfortunately my system is failing to upgrade/reinstall Imagemagick via Homebrew, from source or otherwise…don't know the issue. But glad this option is recorded for others to try. Thanks for that news!

**Comment by Leofystic** [comment]
Can someone bring this to Huggingface demo ? Please. I don't know how to code. I have color depthmap from "Depth Anything v2" But I cannot know how to convert it to grayscale depthmap to use in 3D software. Please help.

**Reply by WorldofDepth** [reply]
FYI this Hugging Face Depth Anything v2 demo already has a grayscale depth map download option: http://huggingface.co/spaces/depth-anything/Depth-Anything-V2
